sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureFileSystemProvider.java

```diff
@@ -363,11 +358,36 @@ BlobContainerClient getContainerClient(BlobClient client) {
     }
 
     /**
+     * As noted by the NIO docs, this method is not atomic. It is possible to delete a file in use by another process,
+     * and doing so will not immediately invalidate any channels open to that file--they will simply start to fail.
+     * Root directories cannot be deleted even when empty.
+     *
      * {@inheritDoc}
      */
     @Override
     public void delete(Path path) throws IOException {
+        // Basic validation. Must be an AzurePath. Cannot be a root.
+        AzurePath aPath = validatePathInstanceType(path);
+        validateNotRoot(path, "Delete");
+
+        // Get client.
+        BlobClient blobClient = aPath.toBlobClient();
+
+        // Check directory status--possibly throw DirectoryNotEmpty or NoSuchFile.
+        DirectoryStatus dirStatus = checkDirStatus(blobClient);
+        if (dirStatus.equals(DirectoryStatus.DOES_NOT_EXIST)) {
+            throw Utility.logError(logger, new NoSuchFileException(path.toString()));
+        }
+        if (dirStatus.equals(DirectoryStatus.NOT_EMPTY)) {
+            throw Utility.logError(logger, new DirectoryNotEmptyException(path.toString()));
+        }
 
+        // After all validation has completed, delete the resource.
+        try {
+            blobClient.delete();
+        } catch (BlobStorageException e) {
+            throw Utility.logError(logger, new IOException(e));
```

Comment:
Do we want to log and throw an exception if the exception is that the blob doesn't exist? I'm good with either, just a theoretical question on what delete on something that doesn't exist means.

Reply:
Won't checkDirStatus do the blob doesn't exist validation for us? It gets the container client and lists paths with path = blob name and returns diff codes depending on the number of elements underneath it

Reply:
I suppose something could also happen between that check and the delete

Reply:
It's true we do check for no such blob based on the dir status earlier and also true that something else could happen in between. I added a specific case for a BlobNotFound error code and throw a NoSuchFileException in that case as well. 

I would prefer to throw in the case of no such file because the docs say we can, so it's more informative and because there's a deleteIfExists api that customers can call if they don't want the exception behavior.